proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,754 @@
+at the transport protocol layer, the resource graph layer is
+essentially irrelevant in that case.
+
+### Dynamic Parameters
+
+With the above property in mind, this document proposes the following
+data structures:
+- **Dynamic parameters**, which are a set of key/value pairs sent by the
+  client when subscribing to a resource.
+- **Dynamic parameter constraints**, which are a set of criteria that
+  can be used to determine whether a set of dynamic parameters matches
+  the constraints.  These constraints are part of the cache key for an
+  xDS resource (in addition to the resource name itself) on xDS servers,
+  xDS clients, and xDS caching proxies.  This provides a mechanism to
+  represent multiple variants of a given resource in a cacheable way.
+
+Both of these data structures are used in the xDS transport protocol,
+but they are not part of the resource name and therefore do not appear as
+part of the resource graph.
+
+When a client subscribes to a resource, it specifies a set of dynamic
+parameters.  In response, the server will send a resource whose dynamic
+parameter constraints match the dynamic parameters in the subscription
+request.  The client will use the dynamic parameter constraints on the
+returned resource to determine which of its subscriptions the resource is
+associated with.
+
+Dynamic parameters, unlike context parameters, will not be
+exact-match-only.  Dynamic parameter constraints will be able to represent
+certain simple types of flexible matching, such as matching an exact
+value or the existance of a key, and simple AND and OR combinations
+of constraints.  This flexible matching semantic means that there may be
+ambiguities when determining which resources match which subscriptions,
+which are discussed below.
+
+#### Constraints Representation
+
+Dynamic parameter constraints will be represented in protobuf form as follows:
+
+```proto
+message DynamicParameterConstraints {
+  // A list of constraints that may be combined with AND or OR semantics.
+  message ConstraintList {
+    // A constraint for a given key.
+    message Constraint {
+      message Exists {}
+      // The key to match against.
+      string key = 1;
+      // How to match.
+      oneof constraint_type {
+        // Matches this exact value.
+        string value = 2;
+        // Key is present (matches any value except for the key being absent).
+        Exists exists = 3;
+      }
+      // If set to true, the match is inverted -- i.e., the key must NOT
+      // match the specified value.
+      bool invert = 4;
+    }
+
+    enum MatchType {
+      // Default value.
+      MATCH_TYPE_UNSPECIFIED = 0;
+      // Logical AND of constraints.
+      MATCH_TYPE_AND = 1;
+      // Logical OR of constraints.
+      MATCH_TYPE_OR = 2;
+    }
+
+    // A list of key/value constraints.
+    repeated Constraint constraints = 1;
+
+    // How to match the constraints.
+    MatchType match_type = 2;
```

Comment:
I think the trade-off here is that the more complex we make this, the more flexible it is, but the harder it is for caches (in both clients and servers) to optimize the matching implementation.  See the section I added below on implementation considerations for some thoughts on this.

That having been said, given that this will be hard to extend later, I do agree that if we think we'll need something, we should make sure to add it now.  We just need to figure out where the right balance is, since we're making a trade-off here.

Reply:
I think flat is OK actually; any tree can be transformed to https://en.wikipedia.org/wiki/Disjunctive_normal_form.

Reply:
@mattklein123 This thread is also related to the other one that I tagged you on.